### PR TITLE
Fix transparent icons in runtime.

### PR DESCRIPTION
**Pull request by tehnick**
But it does not fix transparent icons for WINE applications at startup.

Unfortunately your workaround with QTimer::singleShot() and SNIProxy::update() does not help to WINE applications. For all Wine applications, which I have tested, icons become correct only after (any) mouse click on them in SNI. So another workaround is required....

### CODE REVIEW

**Comment by Crazy-Hopper** [comment]
Creating the container window on-screen, e.g on pixel (0, 0) seems to make wine icons to be drawn immediately. Could you please test it?

File: sniproxy.cpp [hunk 1]
```diff
@@ -153,8 +153,8 @@ SNIProxy::SNIProxy(xcb_window_t wid, QObject* parent):
     const uint32_t windowMoveConfigVals[2] = { 0, 0 };
 
     xcb_configure_window(c, wid,
-                             XCB_CONFIG_WINDOW_X | XCB_CONFIG_WINDOW_Y,
-                             windowMoveConfigVals);
```
Comment (davidedmundson) [inline sniproxy.cpp:157]: it's not good to have a tonne of unrelated changes in a patch

File: sniproxy.cpp [hunk 2]
```diff
@@ -190,7 +190,40 @@ SNIProxy::~SNIProxy()
...
+
+    if (m_pixmap.isNull()) {
+        m_pixmap = QPixmap::fromImage(newImage);
+    }
+    else if (!newImage.isNull()) {
+        auto isTransparentImage = [](const QImage &image) -> bool {
+            unsigned long long sum = 0;
+
+            for (int x = 0; x < image.width(); ++x) {
+                for (int y = 0; y < image.width(); ++y) {
```
Comment (davidedmundson) [inline sniproxy.cpp]: image.height()

File: sniproxy.cpp [hunk 3]
```diff
@@ -190,7 +190,40 @@ SNIProxy::~SNIProxy()
...
+                    sum += qAlpha(image.pixel(x, y));
+                }
+            }
+
+            if (sum < 255) {
```
Comment (davidedmundson) [inline sniproxy.cpp]: it'd be faster if you reversed the logic; 

in the loop break if the sum ever gets bigger than 255

File: sniproxy.cpp [hunk 4]
```diff
@@ -190,7 +190,40 @@ SNIProxy::~SNIProxy()
...
 void SNIProxy::update()
 {
-    m_pixmap = QPixmap::fromImage(getImageNonComposite());
+    QImage newImage = getImageNonComposite();
+
+    if (m_pixmap.isNull()) {
+        m_pixmap = QPixmap::fromImage(newImage);
+    }
+    else if (!newImage.isNull()) {
+        auto isTransparentImage = [](const QImage &image) -> bool {
```
Comment (davidedmundson) [inline sniproxy.cpp]: if you don't capture anything you may as well just make it a normal static method rather than a lambda

File: sniproxy.cpp [hunk 5]
```diff
@@ -190,7 +190,40 @@ SNIProxy::~SNIProxy()
 
 void SNIProxy::update()
 {
-    m_pixmap = QPixmap::fromImage(getImageNonComposite());
+    QImage newImage = getImageNonComposite();
```
Comment (davidedmundson) [inline sniproxy.cpp]: const

File: sniproxy.cpp [hunk 6]
```diff
@@ -190,7 +190,40 @@ SNIProxy::~SNIProxy()
...
+            return false;
+        };
+
+        if (!isTransparentImage(newImage)) {
+            // Update icon only if it is at least partially opaque.
+            // This is just a workaround for X11 bug.
+            m_pixmap = QPixmap::fromImage(newImage);
+        }
+        else {
+            // Delete this before release.
```
Comment (davidedmundson) [inline sniproxy.cpp]: debug is fine, I'll just turn off all debug by default in the Q_LOGGING_CATEGORY line

File: sniproxy.cpp [hunk 7]
```diff
@@ -190,7 +190,40 @@ SNIProxy::~SNIProxy()
...
+                // All pixels more or less tranparent.
+                return true;
+            }
+
+            return false;
+        };
+
+        if (!isTransparentImage(newImage)) {
+            // Update icon only if it is at least partially opaque.
+            // This is just a workaround for X11 bug.
```
Comment (davidedmundson) [inline sniproxy.cpp]: can you give (in the comments in the code) an example of what causes this.

Otherwise I might come up with a new change and think "ooh, maybe I can get rid of this workaround".

Comment (Crazy-Hopper) [inline sniproxy.cpp]: I've seen the wine TeamViewer icon disappear after some arbitrary time. Don't know if this is the case though. But it might be.

**Comment by davidedmundson** [comment]
Thanks, good stuff.

**Comment by Crazy-Hopper** [comment]
Were you referring to the patch or the suggested wine workaround? :-)

File: sniproxy.cpp (hunk 1 above)
Comment (tehnick) [inline sniproxy.cpp:157]: Sorry for this. It was late night...

I'll fix this and other your notes soon.

File: sniproxy.cpp (hunk 7 above)
Comment (tehnick) [inline sniproxy.cpp]: @davidedmundson 
I do not know how to reproduce this manually. It just happens from time to time. See:
https://github.com/davidedmundson/xembed-sni-proxy/issues/3#issuecomment-145302150

File: sniproxy.cpp (hunk 4 above)
Comment (tehnick) [inline sniproxy.cpp]: The main idea was to use good looking inlined function.
I may use simple cycle here without lambda function if you wish.

**Comment by tehnick** [comment]
All remarks were fixed. Please recheck this pull request.

**Comment by tehnick** [comment]
> Were you referring to the patch or the suggested wine workaround? :-)

+1 for this question

**Comment by tehnick** [comment]
@Crazy-Hopper
Do you suggest to create a container window at pixel(0, 0) and then to move it somewhere?

**Comment by davidedmundson** [comment]
If it behaves better with the window on screen.. sure lets do it.

I have another idea I want to try: using _net_wm_sync_request 
http://standards.freedesktop.org/wm-spec/wm-spec-1.4.html#idm139915842324928

It's quite a big change, I'll try and do that tomorrow or so

Another thing to think about; should we delay telling the SNIHost about our StatusNotifierItem until we get an image? Will save us ever posting a blank space.

**Comment by Crazy-Hopper** [comment]
@tehnick I tested without moving it. Just created at (0,0).

**Comment by Crazy-Hopper** [comment]
BUT! there is quite a noticeable flickery at (0,0).
That happens on repeated sendClick().

**Comment by Crazy-Hopper** [comment]
I thought that container window would be moved to where SNI is and will be left there. But apparently something moves it back to (0,0)

**Comment by Crazy-Hopper** [comment]
U-huh. It's the SNIProxy::Scroll() that moves container back to (0,0) :-)

**Comment by tehnick** [comment]
@Crazy-Hopper I may confirm that your workaround works fine.
Now I do not see transparent icons at all. Nor in Wine apps, nor in Gtk+ based apps,
nor in Java-based apps...

**Comment by tehnick** [comment]
@Crazy-Hopper 

> That would explain why you missed that. :-) But definitely not a reason for keeping that code.

Personally I do not use animated icons in system tray since first versions of KDE4, because it has huge memory leaks. Yes, this is old system tray with all its problems. But KDE Plasma 5 has similar (but not the same) [bug](https://bugs.kde.org/show_bug.cgi?id=344879) in a new cool system tray. So it is better to not use animated icons in KDE system tray at all.

As for Psi and Psi+. Just few notes:
- There is special option for notifying user about new unread messages in other way (not by permanently blinking of systray icon).
- Psi and Psi+ are already ported to Qt5. Not sure when package maintainers will build Psi and Psi+ with Qt5 in different Linux-based distros, but you may compile them from sources...

> Well, i reverted in my fork.

Bad idea. Just look at amount of debug messages about skipping transparent icons.

> And I was too quick in judgement. This change does help TeamViewer icon not to disappear completely.

Ok.

> But making it skipping absolutely transparent images (i.e. sum==0) doesn't help Psi either.

Yep, this shouldn't change anything.

> Somewhat improved your patch in e762a4c

Sorry, but it is useless from my point of view, because tray icons are very tiny.

> Gotta search for some better way to solve it.

If I understand correctly xembed system tray in KDE4 don't have this bug because program icon is permanently updated. You just cannot notice such fast blink (only few transparent frames). But such solution is not suitable for SNI.
